tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc

```diff
@@ -410,12 +387,57 @@ void populateHLOToLHLOConversionPattern(MLIRContext* context,
   // clang-format on
 }
 
+/// Removes Lhlo.CopyOp that copies from an allocated buffer to the block
+/// argument. All uses of the buffer are replaced with the block argument.
+struct RedundantCopiesRemoval : mlir::FunctionPass<RedundantCopiesRemoval> {
+  void runOnFunction() override {
+    llvm::SmallVector<mlir::Operation*, 2> eraseList;
+    getFunction().walk([&](mlir::xla_lhlo::CopyOp copyOp) {
+      auto arguments = copyOp.getOperation()->getBlock()->getArguments();
+      if (std::any_of(arguments.begin(), arguments.end(),
+                      [&](mlir::BlockArgument arg) {
+                        return copyOp.output() == arg;
+                      }) &&
+          std::none_of(arguments.begin(), arguments.end(),
+                       [&](mlir::BlockArgument arg) {
+                         return copyOp.operand() == arg;
+                       })) {
+        mlir::Value operand = copyOp.operand();
+        mlir::Value output = copyOp.output();
+        copyOp.erase();
+        for (auto op : operand.getUsers()) {
+          if (!mlir::isa<mlir::DeallocOp>(op)) {
+            op->replaceUsesOfWith(operand, output);
+          }
+        }
+        auto allocOp = operand.getDefiningOp();
+        if (auto deallocOp =
+                mlir::dyn_cast<mlir::DeallocOp>(*allocOp->getUsers().begin())) {
```

Comment:
(transmitting @joker-eph): You are dereferencing the user of the allocOp while you already deleted the copyOp, which mean there might be no user left.

Here is a minimal reproducer that will show it:

func @fusion(%arg0: memref<2x2xf32>) -> memref<2x2xf32> {
  %0 = alloc() {temp = true} : memref<2x2xf32>
  "xla_lhlo.copy"(%0, %arg0) : (memref<2x2xf32>, memref<2x2xf32>) -> ()
  return %0 : memref<2x2xf32>
}

This comes back to my other point about testing: with dedicated testing for this pass, I would have expected this IR to be the first basic test to validate the pass.

Reply:
Thanks for the hint. You are right that this code snippet would break our current version. Although this code could exist from a theoretical point of view as an input program, the current implementation of the `HLO-to-LHLO-Legalization` pass ensures that there will always be a dealloc. Furthermore, it is even worse: currently, the `Std.Return-to-LHLO` converter expects a final dealloc operation to place a proper `CopyOp`. The following code snippet is taken from the `hlo_legalize_to_lhlo.cc` file (starting in line 425):
```c++
if (dealloc == nullptr) {
    returnOp.emitOpError()
        << "Missing dealloc for operand " << operand.index();
    return matchFailure();
}
```
This code snippet also fails in the following tiny test case:
```
func @TestFunc(%arg0: tensor<2xf32>) -> tensor<2xf32> {
   return %arg0 : tensor<2xf32>
}
```